sdk/couchbase-core/src/orphan_reporter.rs

```diff
@@ -0,0 +1,217 @@
+            "last_remote_socket": self.remote_socket,
+        });
+        write!(f, "{}", obj)
+    }
+}
+
+struct OrphanLogJsonEntry {
+    count: u64,
+    top_items: Vec<OrphanLogItem>,
+}
+
+impl Display for OrphanLogJsonEntry {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let bodies: Vec<String> = self.top_items.iter().map(|item| item.to_string()).collect();
+        write!(
+            f,
+            r#"{{"total_count":{},"top_requests":[{}]}}"#,
+            self.count,
+            bodies.join(",")
+        )
+    }
+}
+
+struct OrphanLogService(HashMap<String, OrphanLogJsonEntry>);
+
+impl Display for OrphanLogService {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut parts = Vec::new();
+        for (svc, entry) in &self.0 {
+            parts.push(format!(r#""{}":{}"#, svc, entry));
+        }
+        write!(f, "{{{}}}", parts.join(","))
+    }
+}
+
+// Once we have total_duration added we will order on that
+impl PartialEq for OrphanLogItem {
+    fn eq(&self, other: &Self) -> bool {
+        self.total_server_duration == other.total_server_duration
+    }
+}
+
+impl Ord for OrphanLogItem {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.total_server_duration.cmp(&other.total_server_duration)
+    }
+}
+
+impl PartialOrd for OrphanLogItem {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+struct State {
+    heap: BinaryHeap<Reverse<OrphanLogItem>>,
+    total_count: u64,
+}
+
+// Orphan reporter - Currently only handles/receives KV orphans
+pub struct OrphanReporter {
+    state: Arc<Mutex<State>>,
+    sample_size: usize,
+    reporter_interval: Duration,
+    log_sink: Option<Arc<OrphanSinkFn>>,
+}
+
+impl OrphanReporter {
+    pub fn new(config: OrphanReporterConfig) -> Self {
+        let state = Arc::new(Mutex::new(State {
+            heap: BinaryHeap::new(),
+            total_count: 0,
+        }));
+
+        let log_sink = config.log_sink.clone();
+        let state_clone = state.clone();
+
+        tokio::spawn(async move {
+            trace!(
+                "OrphanReporter started: reporter_interval={:?}, sample_size={}",
+                config.reporter_interval,
+                config.sample_size
+            );
+            let start = Instant::now() + config.reporter_interval;
+            let mut tick = interval_at(start, config.reporter_interval);
+            tick.set_missed_tick_behavior(MissedTickBehavior::Burst);
+
+            loop {
+                tokio::select! {
+                    _ = tick.tick() => {
+                        let mut st = state_clone.lock().unwrap();
+                        if st.total_count == 0 {
+                            continue;
+                        }
+
+                        let obj = Self::create_log_object("kv".to_string(), mem::take(&mut st.heap), st.total_count);
```

Comment:
We might want to pull total_count and heap out of the state before creating the log object, and certainly before logging. As we're using take and total_count is copy anyway we should be able to pull them out and then drop the lock. Locking across the rest of this function is going to be expensive. Something like
```
let (heap, total_count) = {
    let mut st = state_clone.lock().unwrap();
    if st.total_count == 0 {
        continue;
    }

   (mem::take(&mut st.heap), st.total_count)
}
```